### PR TITLE
Support percent mode in number recognizer

### PR DESCRIPTION
**Pull request by STayinloves**
GitHub issue #433 

- Support text ratios in percentage for English
- Support cases like "<fraction number> of" as percentages for English

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.Number/English/Extractors/PercentageExtractor.cs [hunk 1]
```diff
@@ -8,16 +8,30 @@ namespace Microsoft.Recognizers.Text.Number.English
 {
     public sealed class PercentageExtractor : BasePercentageExtractor
     {
-        public PercentageExtractor() : base(NumberExtractor.GetInstance()) { }
+        protected override NumberOptions Options { get; }
+
+
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/English/Extractors/PercentageExtractor.cs]: Nit.

File: .NET/Microsoft.Recognizers.Text.Number/English/Extractors/FractionExtractor.cs [hunk 2]
```diff
@@ -11,54 +10,73 @@ public class FractionExtractor : BaseNumberExtractor
...
+                }
+            };
+
+            if ((Options & NumberOptions.PercentageMode) != 0)
+            {
+                regexes.Add(
+                    new Regex(
+                        NumbersDefinitions.FractionPrepositionWithinPercentModeRegex,
+                        RegexOptions.IgnoreCase | RegexOptions.Singleline)
                     , "FracEng"
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/English/Extractors/FractionExtractor.cs]: Nit: Also move this comma up. As well as the next one. :)

File: .NET/Microsoft.Recognizers.Text.Number/English/Extractors/NumberExtractor.cs [hunk 3]
```diff
@@ -10,29 +10,33 @@ public class NumberExtractor : BaseNumberExtractor
...
-        private static readonly ConcurrentDictionary<string, NumberExtractor> Instances = new ConcurrentDictionary<string, NumberExtractor>();
-
-        public static NumberExtractor GetInstance(NumberMode mode = NumberMode.Default) {
-
-            var placeholder = mode.ToString();
+        private static readonly ConcurrentDictionary<(NumberMode, NumberOptions), NumberExtractor> Instances =
+            new ConcurrentDictionary<(NumberMode, NumberOptions), NumberExtractor>();
 
-            if (!Instances.ContainsKey(placeholder))
+        public static NumberExtractor GetInstance(NumberMode mode = NumberMode.Default, NumberOptions options = NumberOptions.None)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/English/Extractors/NumberExtractor.cs]: Should we define a 'config', like in DateTime?

File: .NET/Microsoft.Recognizers.Text.Number/English/Extractors/PercentageExtractor.cs [hunk 4]
```diff
@@ -8,16 +8,30 @@ namespace Microsoft.Recognizers.Text.Number.English
...
+        public PercentageExtractor(NumberOptions options = NumberOptions.None) : base(
+            NumberExtractor.GetInstance(options: options))
+        {
+            Options = options;
+            Regexes = InitRegexes();
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/English/Extractors/PercentageExtractor.cs]: Why do we need this here now?

File: .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs [hunk 5]
```diff
@@ -129,14 +132,29 @@ protected static ImmutableHashSet<Regex> BuildRegexes(HashSet<string> regexStrs,
...
         {
-            string replaceText = "@" + NumExtType;
+            string replaceNumText = "@" + NumExtType;
+            string replaceFracNumText = "@" + FracNumExtType;
+
             for (int i = 0; i < results.Count; i++)
             {
-                int start = (int)results[i].Start;
-                int end = start + (int)results[i].Length;
+                int start = (int) results[i].Start;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs]: Don't add whitespaces after a cast.

File: .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs [hunk 6]
```diff
@@ -129,14 +132,29 @@ protected static ImmutableHashSet<Regex> BuildRegexes(HashSet<string> regexStrs,
...
+                int end = start + (int) results[i].Length;
                 string str = results[i].Text;
+                var data = new List<(string, ExtractResult)>();
+
+                string replaceText;
+                if (str.Contains(replaceFracNumText) && (Options & NumberOptions.PercentageMode) != 0)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs]: Check options first.

File: .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs [hunk 7]
```diff
@@ -173,33 +206,38 @@ private void PostProcessing(List<ExtractResult> results, string originSource, Di
...
-        private string PreprocessStrWithNumberExtracted(string str, out Dictionary<int, int> positionMap, out IList<ExtractResult> numExtResults)
+        private string PreprocessStrWithNumberExtracted(string str, out Dictionary<int, int> positionMap,
+            out IList<ExtractResult> numExtResults)
         {
             positionMap = new Dictionary<int, int>();
 
             numExtResults = numberExtractor.Extract(str);
             string replaceText = "@" + NumExtType;
+            string replaceFracText = "@" + FracNumExtType;
+            bool fracFlag = (Options & NumberOptions.PercentageMode) != 0;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs]: Please give a better name to fracFlag. Maybe percentModeEnabled.

File: .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs [hunk 8]
```diff
@@ -173,33 +206,38 @@ private void PostProcessing(List<ExtractResult> results, string originSource, Di
...
+                start = (int) extraction.Start;
+                end = (int) extraction.Length + start;
                 for (int j = start; j < end; j++)
                 {
-                    if (match[j] == -1)
+                    if (match[j] == 0)
                     {
-                        match[j] = i;
+                        if (fracFlag && extraction.Data.ToString().StartsWith("Frac"))
+                            match[j] = -(i + 1);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs]: Please always include {}

File: .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs [hunk 9]
```diff
@@ -222,7 +261,7 @@ private string PreprocessStrWithNumberExtracted(string str, out Dictionary<int,
                 start = strPart.Item1;
                 end = strPart.Item2;
                 int type = match[start];
-                if (type == -1)
+                if (type == 0)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs]: What does type means? Should we declare constants for them?

File: .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs [hunk 10]
```diff
@@ -232,11 +271,21 @@ private string PreprocessStrWithNumberExtracted(string str, out Dictionary<int,
                 }
                 else
                 {
-                    string originalText = str.Substring(start, end - start + 1);
-                    ret += replaceText;
-                    for (int i = 0; i < replaceText.Length; i++)
+                    if (type > 0)
+                    {
+                        ret += replaceText;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Extractors/BasePercentageExtractor.cs]: This two conditions only differ by replaceText/replaceFracText.
It's best to assign the right value to a var and use in the code, no need to duplicate the whole thing.

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs [hunk 11]
```diff
@@ -393,7 +397,7 @@ protected ParseResult FracLikeNumberParse(ExtractResult extResult)
                 }
 
                 var intStr = string.Join(" ", fracWords.GetRange(0, mixedIndex));
-                intValue = GetIntValue(GetMatches(intStr));
+                    intValue = GetIntValue(GetMatches(intStr));
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs]: Nit

File: .NET/Microsoft.Recognizers.Text.Number/Parsers/BasePercentageParser.cs [hunk 12]
```diff
@@ -4,27 +4,55 @@ namespace Microsoft.Recognizers.Text.Number
...
-            // do replace text & data from extended info
-            if (extResult.Data is KeyValuePair<string, ExtractResult>)
+            // replace text & data from extended info
+            if (extResult.Data is List<(string, ExtractResult)> extendedData1)
             {
-                var extendedData = (KeyValuePair<string, ExtractResult>)extResult.Data;
-                extResult.Text = extendedData.Key;
-                extResult.Data = extendedData.Value.Data;
-            }
+                if (extendedData1.Count == 2)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Parsers/BasePercentageParser.cs]: Can you add code comments exemplifying in which cases 'if' and 'else' would happen?

File: .NET/Microsoft.Recognizers.Text.Number/Spanish/Extractors/PercentageExtractor.cs [hunk 13]
```diff
@@ -8,7 +8,14 @@ namespace Microsoft.Recognizers.Text.Number.Spanish
 {
     public sealed class PercentageExtractor : BasePercentageExtractor
     {
-        public PercentageExtractor() : base(new NumberExtractor()) { }
+        protected override NumberOptions Options { get; }
+        
+        public PercentageExtractor(NumberOptions options = NumberOptions.None) : base(
+            new NumberExtractor(options: options))
+        {
+            Options = options;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Spanish/Extractors/PercentageExtractor.cs]: Why can't this be in the base constructor already?

File: Specs/Number/English/PercentModelPercentMode.json [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+      "TypeName": "percentage",
+      "Resolution": {
+        "value": "100%"
+      }
+    }
+  ]
+}
+,
+{
+  "Input": "percents of twenty",
```
Comment (tellarin) [inline Specs/Number/English/PercentModelPercentMode.json]: I don't think this is valid English.

File: Specs/Number/English/PercentModelPercentMode.json [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+      "TypeName": "percentage",
+      "Resolution": {
+        "value": "50%"
+      }
+    }
+  ]
+}
+,
+{
+  "Input": "1/4 of",
```
Comment (tellarin) [inline Specs/Number/English/PercentModelPercentMode.json]: Please add the cases without the "of" to make sure they are dealt with correctly.

File: Specs/Number/English/NumberModelPercentMode.json [hunk 16]
```diff
@@ -0,0 +1,43 @@
+[
```
Comment (tellarin) [inline Specs/Number/English/NumberModelPercentMode.json]: Why this file? This is to show that PercentMode if off?

File: Specs/Number/English/NumberModel.json [hunk 17]
```diff
@@ -1658,4 +1658,31 @@
     }
   ]
 }
+,
+{
```
Comment (tellarin) [inline Specs/Number/English/NumberModel.json]: Please add cases for when the "of" will not be tagged as part of the entity if the new mode if not enabled.

File: .NET/Microsoft.Recognizers.Text.Number/English/Extractors/NumberExtractor.cs (hunk 3 above)
Comment (STayinloves) [reply .NET/Microsoft.Recognizers.Text.Number/English/Extractors/NumberExtractor.cs]: For now, we don't have many things to configure, so I think we can add it later.

File: Specs/Number/English/PercentModelPercentMode.json (hunk 14 above)
Comment (STayinloves) [reply Specs/Number/English/PercentModelPercentMode.json]: I copy the tests from origin number model tests.

File: Specs/Number/English/PercentModelPercentMode.json [hunk 18]
```diff
@@ -138,4 +138,10 @@
     }
   ]
 }
+,
+{
+  "Input": "one third",
+  "NotSupported": "javascript,python",
+  "Results": []
```
Comment (tellarin) [inline Specs/Number/English/PercentModelPercentMode.json]: Shouldn't the result here still be 0.3333? I.e. same behaviour if the flag is on or off for this case.

**Pull request merged**